### ISSUE TITLE
storing floating-point values as FLOAT(24) doesn't preserve values over API round trips

### ISSUE THREAD

**Issue by jbrzusto** [MotusWTS/Motus-TO-DO#24]
It appears a bunch of the tag fields are being stored as float (i.e.
MS SQL type FLOAT(24)), e.g. nom_freq and period.

Although that's sufficient precision, the problem is that
code that accesses them is converting them to/from doubles
(i.e. MS SQL type FLOAT(53)).  Most importantly, JSON doesn't support
FLOAT(24), but only FLOAT(53), so some tag parameters are not
preserved by API round-trip:

```
   API (register)       DB storage      API (retrieve)
  166.38 (double) -> 166.38 (float) -> 166.380004882813 (double)
                                    != 166.38 (double)
```
We can code around this, but it would be better to have the back-end
store the full precision of the values we're sending and receiving, and
**we're sending/receiving in FLOAT(53) because that's implicit in using JSON.**

I think the only table fields where it will be worth
using FLOAT(24) are the tag hits table fields:

```
    hits.freq, hits.sigsd, hits.sig, hits.sigsd, hits.noise can all
    remain FLOAT(24), because this is the "innermost" table, with
    a large number of records.
```

It's quite likely I suggested or agreed to non-double column types for
the tag registration database columns back in the mists of time, but
seeing the unanticipated results now, I'd recommend moving to
FLOAT(53) for any column except these.

We'll need to do this carefully to get back to the intended values
after increasing the field precision.  For this, the "sfround" function [here](https://stackoverflow.com/questions/1920933/round-to-n-significant-figures-in-sql)

could be used like so:

```sql
   ALTER TABLE tags ALTER COLUMN nom_freq FLOAT(53);
   UPDATE tags SET nom_freq = sfround(nom_freq, 6);
```

6 significant figures should suffice for all fields which are to be
converted from FLOAT(24) to FLOAT(53).

**Comment by cjardine-bsc** [comment]
I don't think this issue is being caused by FLOAT(24) vs FLOAT(53) as all the columns within that database are currently FLOAT(53) already. Float in SQL server 2008 is 8 byte (SQL uses real to denote 4 byte) . 

I suspect the problem is that the accuracy of the number that is stored will only be precise for the first 8 digits or so. So, the number stored in the database itself is imprecise and therefore should ideally not be used in any equal-to operations. 

API (register)       DB storage      API (retrieve)
  166.38 (double) -> 166.380004882813 (float) -> 166.380004882813 (double)
                                    != 166.38 (double)

But if JSON uses float(53) I'm confused as to where the conversion issue is occurring. 

This provides some insight: https://stackoverflow.com/questions/5098558/float-vs-double-precision

The way around this would be to change the datatype for the fields that need this level of accuracy to decimal, which can be precise and accurate to 38 digits.  Decimal has the drawback of being larger (5 to 17 bytes depending on size) and calculations using decimal tend to be slower than floats.

**Comment by jbrzusto** [comment]
I don't know where the issue is happening, but  all
tag registrations since 2017 Jan 7 are affected,
as well as some others.

### Example of broken floating-point round-trip ###
The registration for tag 25980 was on 2017-06-26 01:14:25
using the API: (post)  https://motus.org/api/tag/register

The left column below is items from the string actually sent
to the API.  The right column is items as returned
later by a call to https://motus.org/data/api/v1.0/searchtags.jsp
with projectID=14 for tag 25980, re-arranged to the same order
as the registration request, but with no changes to item
values.

(param1..3 are blacked out, but were preserved on the round-trip;
also, ts -> tsMotus is a successful round trip with full precision).

Floating point items failing a round-trip are marked with `(**)` between
columns.  The values in the right column are in fact what you'd get if the
value on the left was parsed to an 8-byte double, then converted to a
4-byte float, then converted back to an 8-byte double.  They are **not**
the values you would get by parsing the left value to an 8-byte double
then printing that to 18 decimal digits:
```
  166.38 (text) -> 166.379999999999995 (printing 18 digits of closest 8-byte float)
```
 I don't know where along the way it is happening, but it's somewhere between the tag/register API accepting a JSON string and then the other API returning one.  If all database fields are 
8-byte floats, then the conversion through a 4-byte float is happening before the values
are stored in the DB.
```
{                                 {                                     
                                  "approved":true,                      
"codeSet":"Lotek4",               "codeSet":"Lotek4",                   
                                  "comments":null                       
"date":"20170626011425",                                                
"dateBin":"2017-2",               "dateBin":"2017-2",                   
                                  "deferSec":null,                      
                                  "deployID":null,                      
                                  "elevation":null,                     
"format":"jsonp",                                                       
"hash":"XX",                                                            
"login":"john",                                                         
                                  "latitude":null,                      
                                  "lifeSpan":null,                      
                                  "longitude":null,                     
"manufacturer":"Lotek",           "manufacturer":"Lotek",               
                                  "markerNumber":null,                  
                                  "markerType":null,                    
"mfgID":"283",                    "mfgID":"283",                        
"model":"ANTC-M4-2S"              "model":"ANTC-M4-2S",                 
"nomFreq":166.38,           (**)  "nomFreq":166.3800048828125,          
"offsetFreq":-0.4014,       (**)  "offsetFreq":-0.40139999985694885,    
"param1":XX.XX,                   "param1":XX.XX,   
"param2":XX.XX,                   "param2":XX.XX,   
"param3":XX.XX,                   "param3":XX.XX,   
"param4":0.0,                     "param4":0.0,     
"param5":0.0,                     "param5":0.0,     
"param6":0.0,                     "param6":0.,0     
"paramType":1,                                                          
"period":4.6971,            (**)  "period":4.6971001625061035,          
"periodSD":0.0006,          (**)  "periodSD":6.000000284984708E-4,      
"projectID":14,                   "projectID":14,                       
"pulseLen":2.5,                   "pulseLen":2.5,                       
"pword":"XX",                                                           
"serno":"SG-0815BBBK1352",                                              
                                  "speciesID":null,                     
                                  "status":"pending",                   
                                  "tagID":25980,                        
                                  "tsEnd":null,                         
"ts":1498439593.5037,             "tsSG":1.4984395935037E9,             
                                  "tsStart":null,                       
"type":"ID",                      "type":"ID",                          
}                                 }                                     
```
In case it's useful, here's the response from the tag/register API:
```
{
   "version":"2.0",
   "importID":29643,
   "responseCode":"success-import",
   "tagID":25980,
   "code":"success-terminate"
}
```

**Comment by Hussell** [comment]
prAPIImportTags is using real (4 bytes) instead of float (8 bytes).

**Comment by cjardine-bsc** [comment]
Yup, there it is, the proc was casting the values to real. 

I've just changed them to float (8 bytes). I also changed prAPIImportTagDeployments (deferSec). 

That should do it, I'll leave this open until you have a chance to test to make sure this fixes the issue.

**Comment by jbrzusto** [comment]
Thanks - the next time someone registers a tag we should get confirmation.

**Comment by jbrzusto** [comment]
Can we fix the affected values already in the DB?  The first post in this issue suggests how.
I'm fixing this downstream when I grab the tag registrations, but with the DB in its current state,
there are e.g. tags having the same nominal frequency but appearing not to, so that queries
on the DB might not work as expected.  (a query using e.g. `where nomFreq=166.38`)

**Comment by cjardine-bsc** [comment]
I've updated most of the values, though I did so slightly differently than you suggested as rounding to a predefined number of significant digits would still not allow you to directly match numbers between the database and your tables. 

Interestingly, I found that the LEN() function in SQL will return the length of a float number that was input. 

I.E. 

LEN(0.699999988079071) = 3  Meaning the correct value should be 0.7 
LEN(0.759999990463257) = 4 Meaning the correct value should be 0.76

Which you don't really need to know...but I thought it was neat, so I'm telling you about it. 

I was able to updated all the values for nomFreq, offsetFreq, period_sd and period using a combo of LEN() and ROUND() so they should now all be accurate. Hopefully that works!
